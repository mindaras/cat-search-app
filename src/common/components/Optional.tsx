interface Props {
  if: boolean;
  children: any;
}

const Optional: React.FC<Props> = (props) => {
  if (props.if) {
    return props.children;
  }

  return null;
};

export { Optional };
