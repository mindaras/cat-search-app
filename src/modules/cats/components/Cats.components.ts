import styled from "styled-components";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: calc(
    100vh - ${(p) => p.theme.sizes.search.height} -
      ${(p) => p.theme.sizes.pagination.height}
  );
`;

export { LoaderContainer };
