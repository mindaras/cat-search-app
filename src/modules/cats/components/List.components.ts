import styled from "styled-components";

const Cat = styled.img`
  width: 25%;
  height: calc(100% / 3);
  object-fit: cover;
  border: 4px solid #fff;
  border-radius: 30px;

  ${(p) => p.theme.media.medium} {
    width: 33.33%;
    height: 30vw;
  }

  ${(p) => p.theme.media.small} {
    width: 100%;
    height: 100vw;
  }
`;

interface ContainerProps {
  single: boolean;
}

const Container = styled.div<ContainerProps>`
  position: relative;
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  height: calc(
    100vh - ${(p) => p.theme.sizes.search.height} -
      ${(p) => p.theme.sizes.pagination.height}
  );

  ${(p) => {
    if (p.single) {
      return `
            justify-content: center;
            align-items: center;

            ${Cat} {
                width: 40vw;
                height: 40vw;

                ${p.theme.media.medium} {
                    width: 50vw;
                    height: 50vw;
                }

                ${p.theme.media.small} {
                    width: 80vw;
                    height: 80vw;
                }
            }
        `;
    }
  }};

  ${(p) => p.theme.media.medium} {
    height: auto;
  }
`;

const NoDataText = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  font-size: 24px;
`;

const NoDataContainer = styled.div`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: calc(100vh - ${(p) => p.theme.sizes.search.height});

  ${Cat} {
    opacity: 0.1;
  }

  ${(p) => p.theme.media.small} {
    height: auto;
  }
`;

export { Cat, Container, NoDataContainer, NoDataText };
