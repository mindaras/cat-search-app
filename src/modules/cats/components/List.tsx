import { Optional } from "@common/components/Optional";
import * as Types from "../common/models";
import { Cat, Container, NoDataContainer, NoDataText } from "./List.components";
import pawImage from "@common/images/paw.jpeg";

interface Props {
  cats: Types.Cat[];
}

const List: React.FC<Props> = ({ cats }) => (
  <Container single={cats.length < 2}>
    <Optional if={!!cats.length}>
      {cats.map((cat) => (
        <Cat key={cat.id} src={cat.url} alt="Cat" />
      ))}
    </Optional>
    <Optional if={!cats.length}>
      <NoDataContainer>
        <Cat src={pawImage} alt="Paw" />
        <NoDataText>No data</NoDataText>
      </NoDataContainer>
    </Optional>
  </Container>
);

export { List };
