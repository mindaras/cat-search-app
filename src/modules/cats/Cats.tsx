import { useCallback, useEffect, useState } from "react";
import { LoaderContainer } from "./components/Cats.components";
import { Search } from "@common/components/search/Search";
import * as Types from "./common/models";
import { api, FilterParameters } from "./common/api";
import { Pagination } from "@common/components/pagination/Pagination";
import { Loader } from "@common/components/loader/Loader";
import { Optional } from "@common/components/Optional";
import { Filters, Option } from "@common/models";
import { List } from "./components/List";
import { useSearch } from "./common/hooks/useSearch";

const Cats: React.FC = () => {
  let { searchParams, setSearchParams } = useSearch();
  const [cats, setCats] = useState<Types.Cat[]>([]);
  const [breeds, setBreeds] = useState<Types.Breed[]>([]);
  const [isLoading, setLoading] = useState(false);
  const initialPageCount = 100;
  const [pageCount, setPageCount] = useState(initialPageCount);
  const { page: pageParam, order, breed, search: searchTerm } = searchParams;
  const page = pageParam || "1";
  const breedOptions = [
    { text: "Random breed", value: "" },
    ...breeds.map(({ id, name }) => ({
      text: name,
      value: id,
      selected: breed === id,
    })),
  ];
  const orderOptions = Object.entries(Types.Order).map(([key, value]) => ({
    text: value === Types.Order.Random ? "Random order" : key,
    value,
    selected: order === value,
  }));

  const getBreeds = useCallback(async () => {
    const { data } = await api.getBreeds();
    setBreeds(data);
  }, []);

  useEffect(() => {
    getBreeds();
  }, [getBreeds]);

  const getCat = useCallback(async (searchTerm: string) => {
    const { data } = await api.get({ searchTerm });
    setCats(data);
  }, []);

  useEffect(() => {
    if (!searchTerm) return;
    getCat(searchTerm);
  }, [searchTerm, getCat]);

  const updatePageCount = useCallback(
    (newPageCount: number) => {
      if (order !== Types.Order.Random && newPageCount !== pageCount) {
        setPageCount(newPageCount);
      } else if (
        order === Types.Order.Random &&
        pageCount !== initialPageCount
      ) {
        setPageCount(initialPageCount);
      }
    },
    [pageCount, order]
  );

  const getCats = useCallback(
    async ({ page, order, breed }: FilterParameters) => {
      setLoading(true);

      const { data, pageCount: newPageCount } = await api.getAll({
        page,
        order,
        breed,
      });

      setCats(data);
      updatePageCount(newPageCount);
      setLoading(false);
    },
    [updatePageCount]
  );

  useEffect(() => {
    if (searchTerm) return;
    getCats({ page: parseInt(page) - 1, order, breed });
  }, [page, order, searchTerm, breed, getCats]);

  const onSearchTermChange = (term: string) => {
    setSearchParams({ ...searchParams, search: term });
  };

  const onOrderFilterChange = (option: Option) => {
    setSearchParams({ ...searchParams, order: option.value as string });
  };

  const onBreedFilterChange = (option: Option) => {
    setSearchParams({ ...searchParams, breed: option.value as string });
  };

  const onPageChange = (newPage: number) => {
    const pageParam = (newPage + 1).toString();

    if (pageParam !== page) {
      setSearchParams({ ...searchParams, page: pageParam });
    }
  };

  const filters: Filters = [
    {
      options: orderOptions,
      onChange: onOrderFilterChange,
    },
    {
      options: breedOptions,
      onChange: onBreedFilterChange,
    },
  ];

  return (
    <div>
      <Search
        value={searchTerm}
        filters={filters}
        disabled={isLoading}
        filtersDisabled={!!searchTerm}
        onChange={onSearchTermChange}
      />
      <Optional if={isLoading}>
        <LoaderContainer>
          <Loader />
        </LoaderContainer>
      </Optional>
      <Optional if={!isLoading}>
        <List cats={cats} />
      </Optional>
      <Optional if={cats.length > 1}>
        <Pagination
          page={parseInt(page, 10) - 1}
          total={pageCount}
          disabled={isLoading}
          onChange={onPageChange}
        />
      </Optional>
    </div>
  );
};

export { Cats };
