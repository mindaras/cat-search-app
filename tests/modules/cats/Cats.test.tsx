import renderer from "react-test-renderer";
import { Container } from "@testUtils/components/Container";
import { Cats } from "@modules/cats/Cats";

describe("Cats", () => {
  it("matches a snapshot", () => {
    const component = renderer.create(
      <Container>
        <Cats />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
