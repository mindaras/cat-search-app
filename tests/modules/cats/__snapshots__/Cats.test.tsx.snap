// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cats matches a snapshot 1`] = `
<div>
  <div
    className="sc-gKPRtg jIWpuK"
    disabled={false}
  >
    <input
      className="sc-iBYQkv jYjfUp"
      disabled={false}
      onChange={[Function]}
      placeholder="Image id..."
      type="text"
      value=""
    />
    <div
      className="sc-dkrFOg ifKtkp"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="sc-hLBbgP fNKXvp"
      >
         
        <div
          className="sc-gswNZR fMZMjG"
        >
          <svg
            viewBox="0 0 256 256"
          >
            <g
              transform="translate(1.4065934065934016 1.4065934065934016) scale(2.81 2.81)"
            >
              <path
                d="M 90 24.25 c 0 -0.896 -0.342 -1.792 -1.025 -2.475 c -1.366 -1.367 -3.583 -1.367 -4.949 0 L 45 60.8 L 5.975 21.775 c -1.367 -1.367 -3.583 -1.367 -4.95 0 c -1.366 1.367 -1.366 3.583 0 4.95 l 41.5 41.5 c 1.366 1.367 3.583 1.367 4.949 0 l 41.5 -41.5 C 89.658 26.042 90 25.146 90 24.25 z"
                strokeLinecap="round"
                transform=" matrix(1 0 0 1 0 0) "
              />
            </g>
          </svg>
        </div>
      </div>
      <div
        className="sc-eDvSVe czzLRF"
        open={false}
      >
        <div
          className="sc-jSUZER cbMzIT"
          onClick={[Function]}
        >
          Random order
        </div>
        <div
          className="sc-jSUZER cbMzIT"
          onClick={[Function]}
        >
          Ascending
        </div>
        <div
          className="sc-jSUZER cbMzIT"
          onClick={[Function]}
        >
          Descending
        </div>
      </div>
    </div>
    <div
      className="sc-dkrFOg ifKtkp"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="sc-hLBbgP fNKXvp"
      >
         
        <div
          className="sc-gswNZR fMZMjG"
        >
          <svg
            viewBox="0 0 256 256"
          >
            <g
              transform="translate(1.4065934065934016 1.4065934065934016) scale(2.81 2.81)"
            >
              <path
                d="M 90 24.25 c 0 -0.896 -0.342 -1.792 -1.025 -2.475 c -1.366 -1.367 -3.583 -1.367 -4.949 0 L 45 60.8 L 5.975 21.775 c -1.367 -1.367 -3.583 -1.367 -4.95 0 c -1.366 1.367 -1.366 3.583 0 4.95 l 41.5 41.5 c 1.366 1.367 3.583 1.367 4.949 0 l 41.5 -41.5 C 89.658 26.042 90 25.146 90 24.25 z"
                strokeLinecap="round"
                transform=" matrix(1 0 0 1 0 0) "
              />
            </g>
          </svg>
        </div>
      </div>
      <div
        className="sc-eDvSVe czzLRF"
        open={false}
      >
        <div
          className="sc-jSUZER cbMzIT"
          onClick={[Function]}
        >
          Random breed
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-eDWCr gwNTpN"
  >
    <div
      className="sc-ksBlkl kBZAiz"
    >
      <img
        alt="Paw"
        className="sc-csuSiG ifTmSc"
        src=""
      />
      <div
        className="sc-bqWxrE cjJqQw"
      >
        No data
      </div>
    </div>
  </div>
</div>
`;
